### PR TITLE
Hash table

### PR DESCRIPTION
**Pull request by Vladislav-Sirenko**

### CODE REVIEW

File: DataStructures/HashTable.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+namespace Hash_Table
+{
+    public class HashTable : IHashTable
+    {
+        private readonly LinkedList.LinkedList<KeyValue> _hashtable;
+        private class KeyValue
+        {
+            public int Key { get; private set; }
+            public object Value { get; set; }
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable.cs]: Why not generic?

File: DataStructures/HashTable.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data;
+using LinkedList;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Hash_Table
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable.cs]: Please check out C# naming conventions for namespaces.

File: DataStructures/HashTable.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            set
+            {
+                if (value == null)
+                {
+                    int position = 0;
+                    foreach (var pair in _hashtable)
+                    {
+                        if (pair.Key.Equals(key.GetHashCode()))
+                        {
+                            _hashtable.RemoveAt(position);
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable.cs]: Why do you remove when the value is NULL?

File: DataStructures/HashTable.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data;
+using LinkedList;
+using System.Linq;
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable.cs]: Please remove unnecessary namespaces.

File: DataStructures/HashTable.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Hash_Table
+{
+    public class HashTable : IHashTable
+    {
+        private object[] _array;
+        public int Length => _array.Length;
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable.cs]: This is not quite true. Your hashtable has the length equal to 3 once you instantiate it with no items.

File: DataStructures/HashTable.cs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            _array = new object[3];
+        }
+
+        public int Hash(object key)
+        {
+            return key.GetHashCode() & 0x7fffffff % 10;
+        }
+        public bool Contains(object key)
+        {
+            if (_array[Hash(key)] == null)
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable.cs]: Possible IndexOutOfRangeException.

File: DataStructures/HashTable.cs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+            return true;
+        }
+
+        public void Add(object key, object value)
+        {
+            int index = Hash(key);
+            if (index > _array.Length - 1)
+            {
+                object[] temp = new object[index + 1];
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable.cs]: temp is not the best name here. You can do better :)

File: DataStructures/HashTable.cs [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                for (int i = 0; i <= _array.Length - 1; i++)
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable.cs]: Please use Array.Copy or other capabilities of the standard library.

File: DataStructures/HashTable.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                {
+                    temp[i] = _array[i];
+                }
+
+                _array = temp;
+            }
+
+            if (_array[index] != null)
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable.cs]: Does it make sense to before you resize the array?

File: DataStructures/HashTable.cs [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+{
+    public class HashTable : IHashTable
+    {
+        private object[] _array;
+        public int Length => _array.Length;
+
+
+        public HashTable()
+        {
+            _array = new object[3];
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable.cs]: 3 is a magic number. Avoid this when you can. Instead, you can create a constant variable and give it a name DefaultSize or something like that

File: DataStructures/HashTable.cs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        public object this[object key]
+        {
+            get => _array[Hash(key)];
+
+            set
+            {
+                int index = Hash(key);
+                if (index > _array.Length - 1)
+                {
+                    object[] temp = new object[index + 1];
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable.cs]: See comments of Add method

File: DataStructures/HashTable.cs [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+                }
+
+                _array[index] = value;             
+            }
+        }
+
+        public bool TryGet(object key, out object value)
+        {
+            int index = Hash(key);
+            if (_array[index]==null)
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable.cs]: IndexOutOfRangeException

File: DataStructures/HashTable.cs [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+            if (_array[index] != null)
+            {
+                throw new DuplicateNameException();
+            }
+
+            _array[index] = value;
+        }
+
+        public object this[object key]
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable.cs]: There are bugs as well as the code can be improved. Try to resolve it on your own. Let me know if you need any help.

**Review by evgeniy-glushin — CHANGES_REQUESTED** [review]
I'm okay with the implementation once all the comments are resolved.
Try to create unit tests to reproduce the bugs if you have any. 
I will take a look at unit tests later.

File: DataStructures/HashTable.cs (hunk 4 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable.cs]: Done

File: DataStructures/HashTable.cs (hunk 2 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable.cs]: Understand

File: DataStructures/HashTable.cs (hunk 9 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable.cs]: Yes , because of index may be out of range of array(possible NullReferenceException)

File: DataStructures/HashTable.cs (hunk 10 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable.cs]: Okay

File: DataStructures/HashTable.cs (hunk 12 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable.cs]: Thank you. I didn`t see his case.

File: DataStructures/HashTable.cs (hunk 9 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable.cs]: Yes

File: DataStructures/HashTable/HashTable.cs [hunk 14]
```diff
@@ -36,6 +32,10 @@ public bool Contains(object key)
         public void Add(object key, object value)
         {
             int index = Hash(key);
+            if (index <= _array.Length - 1 && _array[index] != null)
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs]: index <= _array.Length - 1 This is not DuplicateNameException

File: DataStructures/HashTable/HashTable.cs [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+            {
+                int index = Hash(key);
+                if (value == null && index > _array.Length - 1)
+                {
+                    return;
+                }
+                if (index > _array.Length - 1)
+                {
+                    object[] temp = new object[index + 1];
+                    for (int i = 0; i <= _array.Length - 1; i++)
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs]: Do u copy one array to another here? If yes use standard capabilities _array.CopyTo(...).

File: DataStructures/HashTable/HashTable.cs [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        public object this[object key]
+        {
+            get => _array[Hash(key)];
+
+            set
+            {
+                int index = Hash(key);
+                if (value == null && index > _array.Length - 1)
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs]: I think the condition like
 if (index > _array.Length - 1) {}
 else {}
looks simpler. Does it make sense?

File: DataStructures/HashTable/HashTable.cs [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+            int index = Hash(key);
+            if (index > _array.Length - 1)
+            {
+                value = null;
+                return false;
+            }
+
+            if (_array[index] == null)
+            {
+                throw new NullReferenceException();
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs]: return false - looks better to me.

File: DataStructures/HashTable/HashTable.cs [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+                _array[index] = value;
+            }
+        }
+
+        public bool TryGet(object key, out object value)
+        {
+            int index = Hash(key);
+            if (index > _array.Length - 1)
+            {
+                value = null;
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs:100]: Is it necessary to set it to NULL?

File: DataStructures/HashTable/HashTable.cs [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+        {
+            if (_array[Hash(key)] == null)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        public void Add(object key, object value)
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs:34]: Add logic should be shared between the Add method and the indexer.

File: DataStructuresTest/HashTableTests.cs [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+        }
+
+        [TestMethod]
+        public void Getter_when_set_1_item_then_should_return_value()
+        {
+            //arrange
+            var hashTable = new HashTable();
+
+            //act
+            hashTable[1] = "one";
```
Comment (evgeniy-glushin) [inline DataStructuresTest/HashTableTests.cs]: "one" could be a variable

File: DataStructures/HashTable/HashTable.cs (hunk 17 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable/HashTable.cs]: If you try to get an element by key which does not exist in the hash table – throw an exception.
It`s a creteria.
But I can change it to false)

File: DataStructures/HashTable/HashTable.cs (hunk 18 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable/HashTable.cs:100]: The out parameter  must be assigned to before control leaves the current method

File: DataStructures/HashTable/HashTable.cs (hunk 15 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable/HashTable.cs]: Understand

File: DataStructures/HashTable/HashTable.cs (hunk 16 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable/HashTable.cs]: Yes

File: DataStructures/HashTable/HashTable.cs (hunk 19 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable/HashTable.cs:34]: Done

File: DataStructures/HashTable/HashTable.cs (hunk 17 above)
Comment (evgeniy-glushin) [reply DataStructures/HashTable/HashTable.cs]: That's fine then. Leave it as is.

File: DataStructures/HashTable/HashTable.cs (hunk 18 above)
Comment (evgeniy-glushin) [reply DataStructures/HashTable/HashTable.cs:100]: Makes sense

File: DataStructures/HashTable/HashTable.cs [hunk 21]
```diff
@@ -36,40 +36,37 @@ public void Add(object key, object value)
...
-                }
-
-                _array = temp;
+                Copy(out _array, index);
             }
 
             _array[index] = value;
         }
 
+        public void Copy(out object[] array, int index)
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs]: Make it without out keyword, please.

File: DataStructuresTest/HashTableTests.cs [hunk 22]
```diff
@@ -8,6 +8,9 @@ namespace HashTableTests
     [TestClass]
     public class HashTableTests
     {
+        private const string _one = "one";
```
Comment (evgeniy-glushin) [inline DataStructuresTest/HashTableTests.cs]: 1. Check out how constants should be called on the internet. 
2. It's rather AnyItem than _one.

File: DataStructures/HashTable/HashTable.cs (hunk 19 above)
Comment (evgeniy-glushin) [reply DataStructures/HashTable/HashTable.cs:34]: Copy is not enough. Make it similar to the indexer and extract the shared logic.

File: DataStructures/HashTable/HashTable.cs [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+        public void Copy(object[] array, int index)
+        {
+            object[] temp = new object[index + 1];
+            _array.CopyTo(temp, 0);
+            _array = temp;
+        }
+
+        public object this[object key]
+        {
+            get => _array[Hash(key)];
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs]: Possible NullReferanceException. Please cover this and other corner cases with unit tests.

File: DataStructures/HashTable/HashTable.cs [hunk 24]
```diff
@@ -39,17 +39,17 @@ public void Add(object key, object value)
...
             {
-                Copy(out _array, index);
+                Copy(_array, index);
             }
 
             _array[index] = value;
         }
 
-        public void Copy(out object[] array, int index)
+        public void Copy(object[] array, int index)
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs]: Do you really need input array parameter?

File: DataStructures/HashTable/HashTable.cs [hunk 25]
```diff
@@ -32,62 +38,81 @@ public bool Contains(object key)
...
-            _array[index] = value;
+            _array[index] = new ArrayValue();
+            _array[index].Value = value;
         }
 
-        public void Copy(object[] array, int index)
+        private bool CheckForFull(int index)
         {
-            object[] temp = new object[index + 1];
+            if (index > _array.Length)
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs]: index > _array.Length or index >= _array.Length?

File: DataStructures/HashTable/HashTable.cs [hunk 26]
```diff
@@ -32,62 +38,81 @@ public bool Contains(object key)
...
+                _array = Resize(index + 1);
             }
 
-            _array[index] = value;
+            _array[index] = new ArrayValue();
+            _array[index].Value = value;
         }
 
-        public void Copy(object[] array, int index)
+        private bool CheckForFull(int index)
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs]: IsResizeRequired instead of CheckForFull?

Comment (evgeniy-glushin) [reply DataStructures/HashTable/HashTable.cs]: or CheckSize

File: DataStructures/HashTable/HashTable.cs [hunk 27]
```diff
@@ -32,62 +38,81 @@ public bool Contains(object key)
...
+            if (CheckForFull(index))
             {
                 value = null;
                 return false;
             }
 
-            if (_array[index] == null)
+            if (!Contains(key))
             {
                 throw new NullReferenceException();
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs]: I know this is done according to the requirements and it's done right. But it's confusing we have two behaviors when the hashtable doesn't contain an item. Isn't it? Let's not throw but make it like above (return false).

File: DataStructures/HashTable/HashTable.cs [hunk 28]
```diff
@@ -1,18 +1,23 @@
...
+using System.ComponentModel;
 using System.Data;
 
 
 namespace DataStructures.HashTable
 {
     public class HashTable : IHashTable
     {
-        private object[] _array;
+        private class ArrayValue
```
Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs]: Please take a look where Node<> class is in LinkedList (at the end of the file) and do the same with ArrayValue class.

Comment (evgeniy-glushin) [inline DataStructures/HashTable/HashTable.cs]: My suggestion is to rename ArrayValue to HashTableItem or something like that.

File: DataStructures/HashTable/HashTable.cs (hunk 26 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable/HashTable.cs]: Exactly)

File: DataStructures/HashTable/HashTable.cs (hunk 27 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable/HashTable.cs]: Okay

File: DataStructures/HashTable/HashTable.cs (hunk 28 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable/HashTable.cs]: Understand

Comment (Vladislav-Sirenko) [reply DataStructures/HashTable/HashTable.cs]: Okay

File: DataStructures/HashTable/HashTable.cs (hunk 25 above)
Comment (Vladislav-Sirenko) [reply DataStructures/HashTable/HashTable.cs]: It makes sense
